routing/index_graph.cpp

```diff
@@ -1,9 +1,94 @@
+#include "routing/index_graph.hpp"
+
+#include "routing/restrictions_serialization.hpp"
 
 #include "base/assert.hpp"
 #include "base/exception.hpp"
 
-#include "std/limits.hpp"
+#include <algorithm>
+#include <limits>
+
+namespace
+{
+using namespace routing;
+using namespace std;
+
+bool IsRestricted(RestrictionVec const & restrictions, vector<SegmentEdge> const & edges,
+                  Segment const & u, Segment const & v, bool isOutgoing)
+{
+  uint32_t const featureIdFrom = isOutgoing ? u.GetFeatureId() : v.GetFeatureId();
+  uint32_t const featureIdTo = isOutgoing ? v.GetFeatureId() : u.GetFeatureId();
+
+  // Looking for at least one restriction of type No from |featureIdFrom| to |featureIdTo|.
+  if (binary_search(restrictions.cbegin(), restrictions.cend(),
+                    Restriction(Restriction::Type::No, {featureIdFrom, featureIdTo})))
+  {
+    if (featureIdFrom != featureIdTo)
+      return true;
+
+    // @TODO(bykoianko) According to current code if a feature id is marked as a feature with
+    // restrictricted U-turn it's restricted to make a U-turn on the both ends of the feature.
+    // Generally speaking it's wrong. In osm there's information about the end of the feature
+    // where the U-turn is restricted. It's necessary to pass the data to mwm and to use it here.
+    // Please see test LineGraph_RestrictionF1F1No for details.
+    return u.GetSegmentIdx() == v.GetSegmentIdx() && u.IsForward() != v.IsForward();
+  }
+
+  // Taking into account that number of restrictions of type Only starting and
+  // ending with the same feature id is relevantly small it's possible to ignore such cases.
+  if (featureIdFrom == featureIdTo)
+    return false;
+
+  // Looking for a range of restrictins of type Only starting from |featureIdFrom|
+  // and finishing with any feature.
+  auto const range = equal_range(restrictions.cbegin(), restrictions.cend(),
+                                 Restriction(Restriction::Type::Only, {featureIdFrom}),
+                                 [](Restriction const & r1, Restriction const & r2) {
+                                   CHECK(!r1.m_featureIds.empty(), ());
+                                   CHECK(!r2.m_featureIds.empty(), ());
+                                   if (r1.m_type != r2.m_type)
+                                     return r1.m_type < r2.m_type;
+                                   return r1.m_featureIds[0] < r2.m_featureIds[0];
+                                 });
+  auto const lower = range.first;
+  auto const upper = range.second;
+
+  // Checking if there's at least one restrictions of type Only from |featureIdFrom| and
+  // ending with |featureIdTo|. If yes, returns false.
+  if (lower != upper && lower->m_featureIds[1] == featureIdTo)
+    return false;
+
+  // Checking if there's at least one Only starting from |featureIdFrom|.
+  // If not, returns false.
+  if (lower == restrictions.cend() || lower->m_type != Restriction::Type::Only ||
```

Comment:
It's an old code. For the time being only Restriction::Type::No is processed in IndexGraph.